Backport of inverse beta variable in pat::Muons for 10_6_X

cericeci opened the pull request:
#### PR description:

This is a (long overdue) backport of https://github.com/cms-sw/cmssw/pull/28212 to 10_6_X in order to save the inverse beta (timing variable) of pat::muons.  We discussed this-very briefly- this in a reco meeting a couple of months ago. As it's been some time should I raise it again next time?

Our goal would be to have it as part of a possible UL re-miniAOD. Should I also comment on issue https://github.com/cms-sw/cmssw/issues/27889 ?

#### PR validation:

Same tests as the original PR:
No compilation errors or of battery of tests from
runTheMatrix.py -l limited -i all

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Backport of https://github.com/cms-sw/cmssw/pull/28212 to include the feature in reminiAOD of UL

Commit:
Added inverse beta variable+error to pat::Muon

Commit:
Code-checks compliance fixes

Commit:
Code check compliance fixes for the header file

Commit:
Removed unused inverseBeta setters in pat::Muon

Some additional minor naming changes in the muon extra timer info reader

Commit:
Moved clarification comment to variable getter instead of the generic Muon Extra Timer reader

Commit:
Added code checks and tested era modifier applied for miniAOD production

cmsbuild commented:
A new Pull Request was created by  @cericeci (Carlos Erice) for CMSSW_10_6_X.

It involves the following packages:

DataFormats/PatCandidates
PhysicsTools/PatAlgos


@perrotta, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @gouskos, @hatakeyamak, @emilbols, @peruzzim, @seemasharmafnal, @mmarionncern, @ahinzmann, @smoortga, @jdolen, @ferencek, @rovere, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @mariadalfonso, @clelange, @riga, @JyothsnaKomaragiri, @mverzett, @cbernet, @gpetruc, @andrzejnovak this is something you requested to watch as well.
@davidlange6, @silviodonato, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
> Our goal would be to have it as part of a possible UL re-miniAOD. Should I also comment on issue #27889 ?

sure, please. Thank you.

perrotta commented:
backport of #28212

cmsbuild commented:
+1
Tested at: ae74db10ffea433c56cd94c05dbaf8ca569098a2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-32a58d/5413/summary.html
CMSSW: CMSSW_10_6_X_2020-03-27-1100
SCRAM_ARCH: slc7_amd64_gcc700

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-32a58d/5413/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3212324
 - DQMHistoTests: Total failures: 35
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211955
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

santocch commented:
+1

slava77 commented on PhysicsTools/PatAlgos/python/producersLayer1/muonProducer_cfi.py:
```diff
@@ -71,7 +71,9 @@
     ),
-
+    # Read and store combined inverse beta
+    addInverseBeta    = cms.bool(False),  
```
considering that this is a case of simply adding a new variable, I think that it should be fine to just fill it.
This should be acceptable for the "no-change" policy. (the most recent example is in #29276 )

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -33,10 +33,12 @@ def miniAOD_customizeCommon(process):
     run2_muon_2017.toModify( process.patMuons, effectiveAreaVec = [0.0566, 0.0562, 0.0363, 0.0119, 0.0064])
     run2_muon_2018.toModify( process.patMuons, effectiveAreaVec = [0.0566, 0.0562, 0.0363, 0.0119, 0.0064])
     run2_muon_2016.toModify( process.patMuons, mvaTrainingFile = "RecoMuon/MuonIdentification/data/mu_2016_BDTG.weights.xml")
+    run2_miniAOD_devel.toModify(process.patMuons, addInverseBeta = cms.bool(True))    
```
changes in this file should not be necessary considering the comment in muonProducer_cfi.py

slava77 reviewed (commented):
It would probably be easier to review the updates if they didn't also pull in the clang-format changes.
https://github.com/cms-sw/cmssw/compare/f3d19990deb25fb553563d355a5e1f57486e76e5...010a58fcfd4b6e1d4d4a62b321e9365697f388ca?w=1 corresponding to #28212 is much smaller.

Anyways, in case there are more backports coming (IIRC, it was mentioned in the last RECO meeting), perhaps this one simplifies diffing the future updates.

slava77 commented:
@cericeci 
please clarify on the status of updating this PR.
Thank you.

Commit:
Addressed several comments:
-Removed era modifier requirement to include the timing variables (and thus the flag that it activated)
-Included the initialization of timing variables in all the rest of the pat::Muon constructors where it was missing

cmsbuild commented:
Pull request #29324 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cericeci commented:
> @cericeci
> please clarify on the status of updating this PR.
> Thank you.

Sorry, I was mostly offline during this easter. I updated it to remove the miniaod_devel era modifier and the multiple flags related to /needed by it.

Should I also supress the whole clang format changes -I think they got pulled when running git rebase- or is it ok as it is?

slava77 commented:
> Should I also supress the whole clang format changes -I think they got pulled when running git rebase- or is it ok as it is?

I don't have a strong preference.
It seems fine to leave as is.

cmsbuild commented:
+1
Tested at: 366abb2bc271f525261fb4017aa797848f15b02d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-32a58d/5684/summary.html
CMSSW: CMSSW_10_6_X_2020-04-13-1100
SCRAM_ARCH: slc7_amd64_gcc700

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-32a58d/5684/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3212324
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211977
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

slava77 commented on PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc:
```diff
@@ -74,53 +72,63 @@ PATMuonHeavyObjectCache::PATMuonHeavyObjectCache(const edm::ParameterSet& iConfi
+  embedTpfmsMuon_ = iConfig.getParameter<bool>("embedTpfmsMuon");
+  embedDytMuon_ = iConfig.getParameter<bool>("embedDytMuon");
+  // embedding of inverse beta variable information
+  muonTimeExtraToken_ =
+      consumes<edm::ValueMap<reco::MuonTimeExtra>>(iConfig.getParameter<edm::InputTag>("sourceMuonTimeExtra"));
```
please bring this and other parts of this PR in line with the master version
```diff
-  muonTimeExtraToken_ =
-      consumes<edm::ValueMap<reco::MuonTimeExtra>>(iConfig.getParameter<edm::InputTag>("sourceMuonTimeExtra"));
+  addInverseBeta_ = iConfig.getParameter<bool>("addInverseBeta");
+  if (addInverseBeta_) {
+    muonTimeExtraToken_ =
+        consumes<edm::ValueMap<reco::MuonTimeExtra>>(iConfig.getParameter<edm::InputTag>("sourceMuonTimeExtra"));
+  }
```
either the ```addInverseBeta_``` flag should be dropped in the master or re-added here

slava77 commented on PhysicsTools/PatAlgos/plugins/PATMuonProducer.h:
```diff
@@ -162,21 +162,23 @@ namespace pat {
     bool embedTpfmsMuon_;
     /// embed track from DYT muon fit into the muon
     bool embedDytMuon_;
+    /// input tag for reading timing information
```
this part is out of sync with the master as well

slava77 commented on PhysicsTools/PatAlgos/python/producersLayer1/muonProducer_cfi.py:
```diff
@@ -72,6 +72,9 @@
     ),
 
+    # Where to read and store from the timing information
+    sourceMuonTimeExtra = cms.InputTag("muons","combined"), #Use combined info, not only csc or dt
```
this is also out of sync with the master version

slava77 reviewed (commented):
the last update went into another direction wrt the master version and removed the ```addInverseBeta``` switch. 
This appears to be a fine solution for the master, if desired.
Either the master needs an update first or this PR needs an update to re-introduce ```addInverseBeta```

Commit:
Added back inverse beta switch

cmsbuild commented:
Pull request #29324 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

cericeci commented:
> Either the master needs an update first or this PR needs an update to re-introduce `addInverseBeta`

I reintroduced addInverseBeta back again, in fact I checked out the files back from the previous (rebased from master) commit. I was overzealous in removing the effects of the reminiAOD era modifier.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 934752845fa1648350e514a1cea69fe7c22f2d57
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-32a58d/5689/summary.html
CMSSW: CMSSW_10_6_X_2020-04-14-1100
SCRAM_ARCH: slc7_amd64_gcc700

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-32a58d/5689/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3212324
 - DQMHistoTests: Total failures: 15
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211975
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/29324	9347528 
- the backport is now consistent with the master version of #28212 . Since this adds a new variable to the pat::Muon, I think that it should be fine to fill it as in the master version, without additional modifiers to disable filling by default.
- jenkins tests pass and comparisons show no [relevant] differences (old/existing variables are not affected)

silviodonato commented:
do you have comments @santocch?

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_10_6_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_11_1_X is complete. This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
